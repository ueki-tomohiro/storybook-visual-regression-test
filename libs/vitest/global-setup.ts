const setup = () => {
  process.env.TZ = "Asia/Tokyo";
};

export default setup;
