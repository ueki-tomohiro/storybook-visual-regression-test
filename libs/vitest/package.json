{
  "name": "@demo-libs/vitest",
  "version": "0.0.3",
  "private": true,
  "description": "Demo Libs Vitest",
  "license": "MIT",
  "type": "module",
  "main": "index.js",
  "devDependencies": {
    "@juggle/resize-observer": "^3.4.0",
    "@testing-library/react": "^16.0.1",
    "@testing-library/user-event": "^14.5.2",
    "@vitejs/plugin-react": "^4.3.1",
    "@vitejs/plugin-react-swc": "^3.7.0",
    "@vitest/coverage-v8": "^2.0.5",
    "canvas": "^2.11.2",
    "jsdom": "^25.0.0",
    "vite-tsconfig-paths": "^5.0.1",
    "vitest": "^2.0.5"
  },
  "engines": {
    "node": "^20",
    "yarn": ">=1"
  },
  "volta": {
    "node": "20.0.0",
    "yarn": "1.22.19"
  }
}
