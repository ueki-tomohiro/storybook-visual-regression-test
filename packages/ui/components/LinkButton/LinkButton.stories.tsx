import React from "react";

import { LinkButton, LinkButtonProps } from "./LinkButton";

const defaultArgs: LinkButtonProps = {
  variant: "primary",
  children: <>Button</>,
  href: "/",
};

export default {
  title: "components/LinkButton",
  component: LinkButton,
};

export const Default = { args: defaultArgs };
