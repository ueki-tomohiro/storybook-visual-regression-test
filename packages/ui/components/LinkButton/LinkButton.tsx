import Link, { LinkProps } from "next/link";
import React from "react";

import { backgroundColor, Variant } from "../Button/variant";

export type LinkButtonProps = LinkProps &
  React.PropsWithChildren<{
    variant: Variant;
  }>;

export const LinkButton: React.FC<LinkButtonProps> = ({ variant, children, ...props }) => {
  return (
    <Link {...props} className={`mx-auto max-w-3xl text-white ${backgroundColor[variant]} rounded px-4 py-2`}>
      {children}
    </Link>
  );
};
