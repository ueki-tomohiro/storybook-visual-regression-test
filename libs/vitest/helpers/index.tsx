import { composeStories, composeStory } from "@storybook/react";
import type { Args } from "@storybook/types";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { cleanup, render } from "@testing-library/react";
import dayjs from "dayjs";
import noop from "lodash/noop";
import React, { ReactNode } from "react";
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";

const date = dayjs("2020-12-04T10:00:00+09:00").toDate();

export const renderFromStoriesTest = (stories: Parameters<typeof composeStories<any>>[0]) => {
  const componentTitle = stories.default.title as string;
  const defaultArgs = (stories.default && stories.default.args) || {};

  describe(componentTitle, () => {
    Object.entries(composeStories(stories)).forEach(([storyName, Story]: [string, any]) => {
      if (storyName === "__esModule") return;
      if (storyName === "default") return;
      if (storyName.match(/Fake/)) return;

      beforeEach(() => {
        vi.setSystemTime(date);
        Object.defineProperty(HTMLMediaElement.prototype, "muted", {
          set: noop,
        });
        const scrollTo = vi.fn();
        vi.spyOn(window, "scrollTo").mockImplementation(scrollTo);
      });
      afterEach(() => {
        vi.useRealTimers();
        cleanup();
      });

      describe(`"${storyName}" story`, () => {
        test("renders without crashing", () => {
          const { container, unmount } = render(<Story {...defaultArgs} {...(Story.args || {})} />);
          expect(container).toMatchSnapshot();
          unmount();
        });
      });
    });
  });
};

export const interactionTestFromStory = <T extends Args>({
  testName,
  story,
  annotations,
  args,
  callback,
}: {
  testName?: string;
  story: Parameters<typeof composeStory<T>>[0];
  annotations: Parameters<typeof composeStory<T>>[1];
  args?: Partial<T>;
  callback: (args: {
    container: HTMLElement;
    rerender: (ui: React.ReactElement) => void;
    Story: ReturnType<typeof composeStory<T>>;
  }) => Promise<void>;
}) => {
  const Story = composeStory<T>(story, { ...annotations, args: { ...annotations.args, ...args } });

  beforeEach(() => {
    vi.setSystemTime(date);
  });
  afterEach(() => {
    vi.useRealTimers();
    cleanup();
  });

  describe(`"${annotations.title}" story`, () => {
    test(`interaction test ${testName || ""}`, async () => {
      const { container, rerender } = render(<Story {...args} />);
      await callback({ container, rerender, Story });
    });
  });
};

export const createQueryWrapper = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: Infinity,
      },
    },
  });

  const QueryWrapper = ({ children }: { children: ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  return {
    queryClient,
    QueryWrapper,
  };
};
