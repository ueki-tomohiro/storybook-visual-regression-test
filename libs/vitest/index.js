import react from "@vitejs/plugin-react-swc";
import tsconfigPaths from "vite-tsconfig-paths";
import { defineConfig } from "vitest/config";

function stubAssetImport() {
  return {
    name: "stub-asset-import",
    transform(_code, id) {
      if (/(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$/.test(id))
        return {
          code: "export default 'test-file-stub'",
        };
    },
  };
}

/** @type { import('vite').UserConfig } */
export default defineConfig({
  plugins: [react(), tsconfigPaths(), stubAssetImport()],
  test: {
    globalSetup: "@demo-libs/vitest/global-setup.ts",
    root: ".",
    environment: "jsdom",
    setupFiles: ["@demo-libs/vitest/vitest.setup.mts"],
    testTransformMode: {
      ssr: ["**/*"],
    },
    testTimeout: 10000,
    css: { modules: { classNameStrategy: "non-scoped" }, include: /.*/ },
    poolOptions: {
      threads: {
        singleThread: true,
      },
    },
    reporters: ["default", "hanging-process"],
  },
});
