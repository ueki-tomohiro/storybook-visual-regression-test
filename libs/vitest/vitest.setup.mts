import "@testing-library/jest-dom/vitest";

import { configure } from "@testing-library/react";
export * from "./helpers";

configure({
  testIdAttribute: "data-testid",
});
