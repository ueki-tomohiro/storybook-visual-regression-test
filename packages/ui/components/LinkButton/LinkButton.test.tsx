import { renderFromStories } from "@demo-libs/vitest/helpers";
import { describe } from "vitest";

import * as stories from "./LinkButton.stories";

describe(stories.default.title, () => {
  renderFromStories(stories);
});
