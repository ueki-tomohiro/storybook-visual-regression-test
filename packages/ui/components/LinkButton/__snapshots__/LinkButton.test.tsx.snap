// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/LinkButton > components/LinkButton > "Default" story > renders without crashing 1`] = `
<div>
  <a
    class="mx-auto max-w-3xl text-white bg-primary rounded px-4 py-2"
    href="/"
  >
    Button
  </a>
</div>
`;
